### PR TITLE
Home Task 4

### PR DESCRIPTION
**Pull request by avric995**
Finished Home Task 4 - working with Redux Async Thunk.

### CODE REVIEW

File: src/api/axios.js [hunk 1]
```diff
@@ -1,5 +1,13 @@
...
 	baseURL: 'http://localhost:4000',
 });
+
+API.interceptors.request.use(function (request) {
+	const token = localStorage.getItem('token');
+	if (token) {
+		request.headers['Authorization'] = `Bearer ${token}`;
+	}
+	return request;
+});
```
Comment (MariooV) [inline src/api/axios.js:13]: What happens if there is a token and it expires?????????

File: src/components/CourseForm/CourseForm.jsx [hunk 2]
```diff
@@ -1,37 +1,69 @@
...
 	const authors = useSelector(selectAllAuthors);
 
+	const { courseId } = useParams();
+
+	const selectedCourse = useSelector((state) =>
+		selectCourseById(state, courseId)
+	);
+
+	const authorOnCourseList = [];
+	let auList = [];
```
Comment (MariooV) [inline src/components/CourseForm/CourseForm.jsx]: const array when push
let array when =

File: src/components/CourseForm/CourseForm.jsx [hunk 3]
```diff
@@ -1,37 +1,69 @@
...
+	if (selectedCourse) {
+		for (const id of selectedCourse.authors) {
+			let sAuth = authors.find((author) => author.id === id);
+			authorOnCourseList.push(sAuth);
```
Comment (MariooV) [inline src/components/CourseForm/CourseForm.jsx]: what if no find?

File: src/components/CourseForm/CourseForm.jsx [hunk 4]
```diff
@@ -1,37 +1,69 @@
...
+
+	const selectedCourse = useSelector((state) =>
+		selectCourseById(state, courseId)
+	);
+
+	const authorOnCourseList = [];
+	let auList = [];
+	if (selectedCourse) {
+		for (const id of selectedCourse.authors) {
+			let sAuth = authors.find((author) => author.id === id);
```
Comment (MariooV) [inline src/components/CourseForm/CourseForm.jsx]: sAuth === selectedAuthor

We need to take care of the names

File: src/components/CourseForm/CourseForm.jsx [hunk 5]
```diff
@@ -1,37 +1,69 @@
...
+			authorOnCourseList.push(sAuth);
+		}
+
+		auList = authors.filter((authors) => !authorOnCourseList.includes(authors));
```
Comment (MariooV) [inline src/components/CourseForm/CourseForm.jsx]: au as in auuuu 🐺 

authorList!!!

File: src/components/CourseForm/CourseForm.jsx [hunk 6]
```diff
@@ -1,37 +1,69 @@
...
+		selectedCourse ? selectedCourse.title : ''
+	);
+
+	const [description, setDescription] = useState(
+		selectedCourse ? selectedCourse.description : ''
+	);
+
+	useEffect(() => {
+		if (!selectedCourse) setAllAuthors(authors);
+	}, [authors, selectedCourse]);
```
Comment (MariooV) [inline src/components/CourseForm/CourseForm.jsx]: Does useEffect need authors

File: src/components/CourseForm/CourseForm.jsx [hunk 7]
```diff
@@ -111,37 +139,26 @@ const CreateCourse = () => {
...
-		newFormData[fieldName] = fieldValue;
+		const title = event.target.value;
+		setTitle(title);
+	};
 
-		setAddCourse(newFormData);
+	const handleAddCourseDescription = (event) => {
+		event.preventDefault();
+		const description = event.target.value;
+		setDescription(description);
```
Comment (MariooV) [inline src/components/CourseForm/CourseForm.jsx:152]:
```suggestion
		setDescription(event.target.value);
```

File: src/components/Courses/Components/CourseCard/CourseCard.jsx [hunk 8]
```diff
@@ -7,7 +7,7 @@ import findAutor from '../../../../helpers/findAuthors';
 import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
 import { faTrash, faEdit } from '@fortawesome/free-solid-svg-icons';
 import { useDispatch } from 'react-redux';
-import { courseDeleted } from '../../coursesSlice';
+import { deleteCourse } from '../../coursesSlice';
```
Comment (MariooV) [inline src/components/Courses/Components/CourseCard/CourseCard.jsx:10]: 👍

File: src/components/Courses/Components/CourseCard/CourseCard.jsx [hunk 9]
```diff
@@ -50,15 +54,30 @@ const CourseCard = ({
...
+								value={
+									<FontAwesomeIcon
+										icon={faTrash}
+										onClick={() => handleDeleteClick(id)}
+									/>
+								}
+							/>
+						) : (
+							''
+						)}
```
Comment (MariooV) [inline src/components/Courses/Components/CourseCard/CourseCard.jsx:80]:
```suggestion
						{logedUser.role === 'admin' ? (
							<>
							    <Button
								    value={
									    <FontAwesomeIcon
										    icon={faEdit}
										    onClick={(e) => handleEditClick(id)}
									    />
								    }
							    />
							    <Button
								    value={
									    <FontAwesomeIcon
										    icon={faTrash}
										    onClick={() => handleDeleteClick(id)}
									    />
								    }
							    />
							</>
						) : (
							''
						)}
						```

File: src/components/Courses/Courses.jsx [hunk 10]
```diff
@@ -76,9 +85,13 @@ const Courses = () => {
...
-					<Link to='add'>
-						<Button value='Add new course' />
-					</Link>
+					{logedUser.role === 'admin' ? (
+						<Link to='add'>
+							<Button value='Add new course' />
+						</Link>
+					) : (
+						''
+					)}
```
Comment (MariooV) [inline src/components/Courses/Courses.jsx:94]:
```suggestion
					{logedUser.role === 'admin' ? (
						<Link to='add'>
							<Button value='Add new course' />
						</Link>
					) : (
						<></>
					)}
```

File: src/components/Courses/coursesSlice.js [hunk 11]
```diff
@@ -11,33 +11,52 @@ const initialState = {
...
+		}
+	}
+);
+
+export const deleteCourse = createAsyncThunk(
+	'courses/deleteCourse',
+	async (id) => {
+		try {
+			const { data } = await API.delete(`${routes.deleteCourse}/${id}`);
+			if (data?.successful === true) return id;
```
Comment (MariooV) [inline src/components/Courses/coursesSlice.js]: Do we need this if we have error handling?

File: src/constants.js [hunk 12]
```diff
@@ -57,5 +57,10 @@ export const routes = {
 	register: '/register',
 	allCourses: '/courses/all',
 	addCourse: '/courses/add',
+	updateCourse: '/courses',
+	deleteCourse: '/courses',
```
Comment (MariooV) [inline src/constants.js]:
```suggestion
	modifyCourse: '/courses',
```

File: src/features/authors/authorsSlice.js [hunk 13]
```diff
@@ -9,28 +9,36 @@ const initialState = {
...
 );
 
+export const addNewAuthor = createAsyncThunk(
+	'courses/addNewAuthor',
+	async (initialAuthor) => {
+		const { data } = await API.post(routes.addAuthor, initialAuthor);
+		return data.result;
+	}
+);
+
```
Comment (MariooV) [inline src/features/authors/authorsSlice.js]: Error Handling?

File: src/features/user/userSlice.js [hunk 14]
```diff
@@ -1,9 +1,31 @@
...
 };
 
+export const logoutUser = createAsyncThunk(
+	'courses/logoutUser',
+	async (token) => {
+		try {
+			const response = await API.delete(routes.logout, {
+				headers: { Authorization: token },
+			});
+			if (response.status === 200) return token;
```
Comment (MariooV) [inline src/features/user/userSlice.js]: Do we need this if we have error handling?

File: src/helpers/UserProtectedRoute.jsx [hunk 15]
```diff
@@ -0,0 +1,13 @@
+import { Navigate } from 'react-router-dom';
+
+import { useSelector } from 'react-redux';
+import { user } from '../features/user/userSlice';
+const UserProtectedRoute = ({ children }) => {
+	const logedUser = useSelector(user);
+
+	return (
+		<>{logedUser.role === 'admin' ? children : <Navigate to={'/courses'} />}</>
```
Comment (MariooV) [inline src/helpers/UserProtectedRoute.jsx]: Do we need <></> here?

**Review by MariooV — APPROVED** [review]
:shipit:

File: src/api/axios.js (hunk 1 above)
Comment (avric995) [reply src/api/axios.js:13]: Who knows 😄

**Pull request merged**